### PR TITLE
Properly keep authority over attached player position server-side

### PR DESCRIPTION
**Pull request by SmallJoker**
See #10908

## To do

This PR is a Work in Progress

## How to test

1. Create a server
2. Check whether you can see other players moving
3. Place cart onto rail, attach
4. [this script](https://github.com/minetest/minetest/issues/10908#issuecomment-774481432)

On low latency connections there are now less anticheat false-positives, but I'm afraid the problem will persist with higher ping. Suggestions to improve this are welcome.

### CODE REVIEW

**Comment by TheTermos** [comment]
I think the server should have full authority over attached player whereabouts so anticheat shouldn't even engage.
If we're lucky it's already the case that server has the authority, so just don't run anticheat if isAttached.
If it's not the case however, that is the client can modify player_sao state directly while attached , then it's a bug and anticheat being disabled will hopefully help track it down.

**Comment by sfan5** [comment]
As with many things movement in MT the client has the authority and the server makes some best-effort checks that the values are sane.
If it was about "simple" attachments, moving authority server-side wouldn't be a problem. When attachment rotation and bones come into play you'd need the server to run model/bone animation to correctly determine where the client should be.

**Comment by TheTermos** [comment]
> When attachment rotation and bones come into play you'd need the server to run model/bone animation to correctly determine where the client should be.

Are you sure? I can't think of any reasons for doing so.
From the perspective of the server attachment position should be equal to parent position, full stop. There's no need for the server to care about bones and rotations.
Position is also the only parameter anticheat cares about. 

Chances are the only thing left to do is to make sure server doesn't accept client's position updates while player is attached.

**Comment by lhofhansl** [comment]
Looks good. I'm going to try today.

As for server vs client. Perhaps this can be improved, but I just imagine a setup where player movement is as laggy as (non player) entities. What do we do when the server is gone into lag-nirvana for a second or two? The player still has to be able to move. Hence client and server will always be out of sync in a sense.

Aside from the fact the server should not be laggy - and that in any case any lag should be disengaged from any time and movement management... Maybe I'm missing something simple?

**Comment by lhofhansl** [comment]
In the extreme case I tried: MineClone2, Number of EmergeThreads set to 0 (which will 14 on my machine), and the Hot-Air-Balloon mod - hacked to allow faster speeds) I see a very slight improvement, it's still very choppy and resets the player-position all the time. This is due to the server going away for multiple seconds generating the terrain in Lua.

**Comment by TheTermos** [comment]
> What do we do when the server is gone into lag-nirvana for a second or two? The player still has to be able to move.

No, when attached, player doesn't have to be able to move during lag, nor should the player be able to move if there's no lag, and he never has been. What moves is the parent entity, and that happens server side. 
Player can't ever control the entity directly, it may move based on keystrokes sent to the server.
Of course during lag, client side the entity can still appear to move based on last received parameters, and the player with it.

@SmallJoker this is probably a fat misunderstanding, or in case it's me who doesn't understand something, can you describe the mechanics of client cheating being possible while player is attached?

**Comment by TheTermos** [comment]
> I tried: [...] Hot-Air-Balloon mod - hacked to allow faster speeds

It's a good idea not to use real mods, some problems may come from the mod, in fact this one seems to work for me but throws a bunch of errors on startup. To test this properly create a simple entity without any logic, then attach the player to it and give it constant velocity.

> At least I think I see a change. It is not obviously better.

My bet is on the placebo effect.

**Comment by SmallJoker** [comment]
@lhofhansl Indeed, it's far from perfect and I hoped to find solutions in the comments here. Lag will always cause problems with this asynchronous position send & checking.
I wonder whether it wouldn't be better to simply stop sending player positions and ignoring them server-side. There are still range checks that would ensure that the client cannot access chests from too far or dig nodes where they cannot reach. This very same problem would occur as well, but less noticeable.

> @SmallJoker this is probably a fat misunderstanding, or in case it's me who doesn't understand something, can you describe the mechanics of client cheating being possible while player is attached?

@TheTermos When the client ignores the attachment on purpose and freely moves to interact with the world. Luckily there are range-checks that sort of limit it.

**Comment by sfan5** [comment]
> I wonder whether it wouldn't be better to simply stop sending player positions and ignoring them server-side.

This hides but doesn't solve the issue. Other anticheat routines rely on the player position being accurate, but it won't be because `move_to` just instantly sets the position to the destination.
Also as I've mentioned before, the server cannot account for attaching to bones (rotations are doable, just extra work).

**Comment by TheTermos** [comment]
> @TheTermos When the client ignores the attachment on purpose and freely moves to interact with the world. Luckily there are range-checks that sort of limit it.

Here's your bug then, if you're worried about interactions then check the distance of interaction server side.

Checking the distance from player_sao to the parent does exactly nothing, because the position is copied form the parent anyways, it doesn't say anything about what happens client side. The best thing anticheat can do is to trigger a position update, which there's no reason for a hacked client to care about.
Unluckily your range-checks do nothing but harm.

**Comment by TheTermos** [comment]
> As of 5.3.0 once attached to some object, the client can just send "I moved to (28000,200,-28000)" and the server will happily accept and save the position.

If this is true, then you've found the fat bug that should be fixed asap.
But are you sure? I've only peeked at the code but I can recall methods that affect position return immediately if isAttached.
Can you identify the part of the code that causes server to accept a position update when player is attached?

**Comment by sfan5** [comment]
> because the position is copied form the parent anyways

In simple cases yes such as the one tested here that is true, but this does not hold in general.

Once again: When attaching, the child object can be attached to a bone of the parent object, this bone will be offset from the origin of the object (also depending on animation state), thus child pos != parent pos which directly contradicts your claim.

> But are you sure?

https://github.com/minetest/minetest/blob/b28749057a614075c1f9ba3f96bb86a6e248a210/src/network/serverpackethandler.cpp#L490-L492

<b>🠗🠗</b>

https://github.com/minetest/minetest/blob/45ccfe26fb6e0a130e4925ec362cccb1f045a829/src/server/player_sao.cpp#L321-L327

<b>🠗🠗</b>

https://github.com/minetest/minetest/blob/4c76239818f5159314f30883f98b977d30aaa26c/src/server/serveractiveobject.h#L86

**Comment by TheTermos** [comment]
> When attaching, the child object can be attached to a bone of the parent object, this bone will be offset from the origin of the object (also depending on animation state), thus child pos != parent pos which directly contradicts your claim.

Fair enough.
But then comes PlayerSAO::step() which seems to set pos = parent pos unconditionally, which would in turn directly contradict yours. ;)
https://github.com/minetest/minetest/blob/b28749057a614075c1f9ba3f96bb86a6e248a210/src/server/player_sao.cpp#L259-L267

@sfan5 @SmallJoker  - looks like there's a good chance that player attachment via armature is non-functional atm, and even if it somewhat works, there's a good chance it hasn't been used yet,
that would create a fortunate opportunity to make it right:

- remove anticheat for attached player, it's useless and can't possibly work in any scenario.
- bring back server authority for attached player scenario by fixing Server::process_PlayerPos() so it doesn't accept pos updates from attached players.

At a later date, if bone based attachments can be made work while preserving server authority, then great, that's good to have.

**Comment by SmallJoker** [comment]
> that would create a fortunate opportunity to make it right:

Yes, that's exactly that I was talking in my previous message. The most recent commit now implements this.

**Comment by TheTermos** [comment]
> However, now the player can again freely move to wherever they have map data

This is okay, it's unavoidable anyway, before ass well a hacked client would just ignore server pos updates.
Let's hope this one works out and there are no further surprises in the code.

Server authority is the best anticheat, and probably the only one viable in oss.
Hope we'll have a chat about going full server authority, maybe after 5.4 is released.

**Comment by sfan5** [comment]
> But then comes PlayerSAO::step() which seems to set pos = parent pos unconditionally, which would in turn directly contradict yours. ;)

Huh that's interesting, I have to admit I didn't check the client code regarding this.
Maybe that means bone attachments are nonfunctional if the child is a player, because if objects were broken too I'm sure someoen would have complained.

File: src/network/serverpackethandler.cpp [hunk 1]
```diff
@@ -894,6 +898,24 @@ bool Server::checkInteractDistance(RemotePlayer *player, const f32 d, const std:
...
+		int parent_id;
+		{
+			std::string bone;
+			v3f attachment_rot;
+			bool force_visible;
+			sao->getAttachment(&parent_id, &bone, &attachment_pos, &attachment_rot, &force_visible);
+		}
+		 if (parent_id) {
+			 // When attached, add the attachment radius to the allowed interaction range
+			 max_d += attachment_pos.getLength();
```
Comment (sfan5) [inline src/network/serverpackethandler.cpp:913]: so does the base position not include the correct attachment offset?

Comment (SmallJoker) [reply src/network/serverpackethandler.cpp:913]: The attachment offset consists of a transformation matrix which is currently not applied/calculated server-side. That would also require to synchronize auto-rotating items which currently just spin "randomly". I think it's good enough to simplify this as a spherical tolerance.

Comment (sfan5) [reply src/network/serverpackethandler.cpp:913]: I agree, was just wondering.

**Comment by TheTermos** [comment]
I wonder if this fixes the eternal problem of 2nd level attachment player self detaching.

I can't think of any other causes than client and server fighting over player's position, so if the problem persists it's probably a good idea to do some more digging.

**Comment by SmallJoker** [comment]
For now I'd rather like to know what's up with Clang. The headers are included. Does it need some special treatment?

**Comment by TheTermos** [comment]
I'm not familiar with Clang, but I hear in case of templates some compilers require both declaration and definition in the same place, maybe try to put it all in the header?

I wonder about something else, what's that smooth translator doing server side, is it needed there for anything in particular?

**Comment by TheTermos** [comment]
How about keeping it purely on the bugfix basis? Improvements such as smooth translators can wait.
Time's becoming of the essence.

**Comment by sfan5** [comment]
I looked into this a bit more and it turns out what @TheTermos said is entirely correct: **The server has and has always had authority of the attached position.**

The code that resets the position every step when attached was added in https://github.com/minetest/minetest/commit/ba3fd63e29252987d4904ec3c120bce4b6758160 (8y ago), so entirely intended the entire time.  The client still sending its' position sounds is merely an oversight, albeit a big one.

This raises a few questions / observations:
* Attachment position, rotation and bone are 100% client-side ("for looks only"). From the server POV attached children are always in the origin of the parent.
* Conversely, the server doesn't tell clients where attached objects are. Clients decide where those should be themselves taking offset, rotation, bone, animation, ... into account.
* ⬆ Why hasn't the obvious mismatch between where the object is visually and where the server thinks it is been a problem yet?
* How could the exploit happen in the first place if the server has the authority?

For the last one I have an answer:
The requirements for this exploit are only fulfilled since e4031156f13d062b03960e7ef1c9221f749f884b (0.4.15). Since then, Interaction packets contain the player position/movement information too which is processed immediately before performing the action.
Thus, a client can choose to teleport anywhere immediately before they perform an action. The server still resets their position in the very next step but the interaction has already happened.

<hr />

For this PR the conclusion is that we don't need `SmoothTranslator` on the server. All we need is to the server to ignore `PlayerPos` when attached and the client to stop sending it too.

**Pull request closed without merging**